File: plugins/plugin-linear/python/src/elizaos_plugin_linear/services/linear.py
import uuid
from datetime import datetime
from typing import Protocol

import httpx

from elizaos_plugin_linear.types import (
    CommentData,
    IssueData,
    LabelData,
    LinearActivityItem,
    LinearAPIError,
    LinearAuthenticationError,
    LinearCommentInput,
    LinearConfig,
    LinearIssueInput,
    LinearSearchFilters,
    ProjectData,
    StateData,
    TeamData,
    UserData,
)


class RuntimeProtocol(Protocol):
    def get_setting(self, key: str) -> str | None: ...


VIEWER_QUERY = """
query Viewer {
    viewer {
        id
        email
        name
    }
}
"""

TEAMS_QUERY = """
query Teams {
    teams {
        nodes {
            id
            name
            key
            description
        }
    }
}
"""

TEAM_QUERY = """
query Team($id: String!) {
    team(id: $id) {
        id
        name
        key
        description
    }
}
"""

USERS_QUERY = """
query Users {
    users {
        nodes {
            id
            name
            email
        }
    }
}
"""

ISSUE_QUERY = """
query Issue($id: String!) {
    issue(id: $id) {
        id
        identifier
        title
        description
        priority
        priorityLabel
        url
        createdAt
        updatedAt
        dueDate
        estimate
        assignee {
            id
            name
            email
        }
        state {
            id
            name
            type
            color
        }
        team {
            id
            name
            key
        }
        labels {
            nodes {
                id
                name
                color
            }
        }
        project {
            id
            name
            description
        }
    }
}
"""

ISSUES_QUERY = """
query Issues($first: Int, $filter: IssueFilter) {
    issues(first: $first, filter: $filter) {
        nodes {
            id
            identifier
            title
            description
            priority
            priorityLabel
            url
            createdAt
            updatedAt
            assignee {
                id
                name
                email
            }
            state {
                id
                name
                type
            }
            team {
                id
                name
                key
            }
        }
    }
}
"""

CREATE_ISSUE_MUTATION = """
mutation CreateIssue($input: IssueCreateInput!) {
    issueCreate(input: $input) {
        success
        issue {
            id
            identifier
            title
            url
        }
    }
}
"""

UPDATE_ISSUE_MUTATION = """
mutation UpdateIssue($id: String!, $input: IssueUpdateInput!) {
    issueUpdate(id: $id, input: $input) {
        success
        issue {
            id
            identifier
            title
            url
        }
    }
}
"""

ARCHIVE_ISSUE_MUTATION = """
mutation ArchiveIssue($id: String!) {
    issueArchive(id: $id) {
        success
    }
}
"""

CREATE_COMMENT_MUTATION = """
mutation CreateComment($input: CommentCreateInput!) {
    commentCreate(input: $input) {
        success
        comment {
            id
            body
            createdAt
        }
    }
}
"""

PROJECTS_QUERY = """
query Projects($first: Int) {
    projects(first: $first) {
        nodes {
            id
            name
            description
            state
            progress
            startDate
            targetDate
            url
            teams {
                nodes {
                    id
                    name
                    key
                }
            }
            lead {
                id
                name
                email
            }
        }
    }
}
"""

LABELS_QUERY = """
query Labels($first: Int, $filter: IssueLabelFilter) {
    issueLabels(first: $first, filter: $filter) {
        nodes {
            id
            name
            color
        }
    }
}
"""

WORKFLOW_STATES_QUERY = """
query WorkflowStates($filter: WorkflowStateFilter) {
    workflowStates(filter: $filter) {
        nodes {
            id
            name
            type
            color
        }
    }
}
"""


class LinearService:
    service_type = "linear"
    capability_description = (
        "Linear API integration for issue tracking, project management, and team collaboration"
    )

    def __init__(self, runtime: RuntimeProtocol) -> None:
        api_key = runtime.get_setting("LINEAR_API_KEY")
        if not api_key:
            raise LinearAuthenticationError("Linear API key is required")

        workspace_id = runtime.get_setting("LINEAR_WORKSPACE_ID")
        default_team_key = runtime.get_setting("LINEAR_DEFAULT_TEAM_KEY")

        self.config = LinearConfig(
            api_key=api_key,
            workspace_id=workspace_id,
            default_team_key=default_team_key,
        )

        self._client = httpx.AsyncClient(
            base_url="https://api.linear.app/graphql",
            headers={
                "Authorization": api_key,
                "Content-Type": "application/json",
            },
            timeout=30.0,
        )

        self._activity_log: list[LinearActivityItem] = []

    @classmethod
    async def start(cls, runtime: RuntimeProtocol) -> "LinearService":
        service = cls(runtime)
        await service._validate_connection()
        return service

    async def stop(self) -> None:
        self._activity_log = []
        await self._client.aclose()

    async def _validate_connection(self) -> None:
        try:
            await self.get_current_user()
        except Exception as e:
            raise LinearAuthenticationError(f"Failed to authenticate with Linear API: {e}") from e

    async def _execute_query(
        self,
        query: str,
        variables: dict[str, str | int | float | bool | list | dict | None] | None = None,
    ) -> dict[str, str | int | float | bool | list | dict | None]:
        payload: dict[str, str | dict] = {"query": query}
        if variables:
            payload["variables"] = variables

        response = await self._client.post("", json=payload)

        if response.status_code == 401:
            raise LinearAuthenticationError("Invalid API key")
        if response.status_code == 429:
            raise LinearAPIError("Rate limit exceeded", status=429)
        if response.status_code >= 400:
            raise LinearAPIError(f"API error: {response.text}", status=response.status_code)

        data = response.json()
        if "errors" in data:
            error_msg = data["errors"][0].get("message", "Unknown error")
            raise LinearAPIError(f"GraphQL error: {error_msg}")

        return data.get("data", {})

    def _log_activity(
        self,
        action: str,
        resource_type: str,
        resource_id: str,
        details: dict[str, str | int | float | bool | None],
        success: bool,
        error: str | None = None,
    ) -> None:
        activity = LinearActivityItem(
            id=f"{int(datetime.now().timestamp() * 1000)}-{uuid.uuid4().hex[:9]}",
            timestamp=datetime.now().isoformat(),
            action=action,
            resource_type=resource_type,  # type: ignore
            resource_id=resource_id,
            details=details,
            success=success,
            error=error,
        )

        self._activity_log.append(activity)

        if len(self._activity_log) > 1000:
            self._activity_log = self._activity_log[-1000:]

    def get_activity_log(
        self,
        limit: int = 100,
        filter_by: dict[str, object] | None = None,
    ) -> list[LinearActivityItem]:
        filtered = list(self._activity_log)

        if filter_by:
            for key, value in filter_by.items():
                filtered = [item for item in filtered if getattr(item, key, None) == value]

        return filtered[-limit:]

    def clear_activity_log(self) -> None:
        self._activity_log = []

    async def get_teams(self) -> list[TeamData]:
        try:
            data = await self._execute_query(TEAMS_QUERY)
            teams = data.get("teams", {}).get("nodes", [])
            self._log_activity("list_teams", "team", "all", {"count": len(teams)}, True)
            return teams  # type: ignore
        except Exception as e:
            self._log_activity("list_teams", "team", "all", {}, False, str(e))
            raise LinearAPIError(f"Failed to fetch teams: {e}")

    async def get_team(self, team_id: str) -> TeamData:
        try:
            data = await self._execute_query(TEAM_QUERY, {"id": team_id})
            team = data.get("team", {})
            self._log_activity("get_team", "team", team_id, {"name": team.get("name")}, True)
            return team  # type: ignore
        except Exception as e:
            self._log_activity("get_team", "team", team_id, {}, False, str(e))
            raise LinearAPIError(f"Failed to fetch team: {e}")

    async def create_issue(self, input_data: LinearIssueInput) -> IssueData:
        try:
            variables = {
                "input": {
                    "title": input_data.title,
                    "teamId": input_data.team_id,
                }
            }

            if input_data.description:
                variables["input"]["description"] = input_data.description
            if input_data.priority is not None:
                variables["input"]["priority"] = input_data.priority
            if input_data.assignee_id:
                variables["input"]["assigneeId"] = input_data.assignee_id
            if input_data.label_ids:
                variables["input"]["labelIds"] = input_data.label_ids
            if input_data.project_id:
                variables["input"]["projectId"] = input_data.project_id
            if input_data.state_id:
                variables["input"]["stateId"] = input_data.state_id
            if input_data.estimate is not None:
                variables["input"]["estimate"] = input_data.estimate
            if input_data.due_date:
                variables["input"]["dueDate"] = input_data.due_date.isoformat()

            data = await self._execute_query(CREATE_ISSUE_MUTATION, variables)
            result = data.get("issueCreate", {})

            if not result.get("success"):
                raise LinearAPIError("Failed to create issue")

            issue = result.get("issue", {})
            self._log_activity(
                "create_issue",
                "issue",
                issue.get("id", "new"),
                {"title": input_data.title, "teamId": input_data.team_id},
                True,
            )

            return issue  # type: ignore
        except Exception as e:
            self._log_activity(
                "create_issue", "issue", "new", {"title": input_data.title}, False, str(e)
            )
            raise LinearAPIError(f"Failed to create issue: {e}")

    async def get_issue(self, issue_id: str) -> IssueData:
        try:
            data = await self._execute_query(ISSUE_QUERY, {"id": issue_id})
            issue = data.get("issue", {})
            self._log_activity(
                "get_issue",
                "issue",
                issue_id,
                {"title": issue.get("title"), "identifier": issue.get("identifier")},
                True,
            )
            return issue  # type: ignore
        except Exception as e:
            self._log_activity("get_issue", "issue", issue_id, {}, False, str(e))
            raise LinearAPIError(f"Failed to fetch issue: {e}")

    async def update_issue(
        self,
        issue_id: str,
        updates: dict[str, str | int | float | bool | list[str] | None],
    ) -> IssueData:
        try:
            variables = {"id": issue_id, "input": updates}
            data = await self._execute_query(UPDATE_ISSUE_MUTATION, variables)
            result = data.get("issueUpdate", {})

            if not result.get("success"):
                raise LinearAPIError("Failed to update issue")

            issue = result.get("issue", {})
            self._log_activity("update_issue", "issue", issue_id, updates, True)
            return issue  # type: ignore
        except Exception as e:
            self._log_activity("update_issue", "issue", issue_id, updates, False, str(e))
            raise LinearAPIError(f"Failed to update issue: {e}")

    async def delete_issue(self, issue_id: str) -> None:
        try:
            data = await self._execute_query(ARCHIVE_ISSUE_MUTATION, {"id": issue_id})
            result = data.get("issueArchive", {})

            if not result.get("success"):
                raise LinearAPIError("Failed to archive issue")

            self._log_activity("delete_issue", "issue", issue_id, {"action": "archived"}, True)
        except Exception as e:
            self._log_activity(
                "delete_issue", "issue", issue_id, {"action": "archive_failed"}, False, str(e)
            )
            raise LinearAPIError(f"Failed to archive issue: {e}")

    async def search_issues(self, filters: LinearSearchFilters) -> list[IssueData]:
        try:
            filter_obj: dict[str, str | int | list | dict] = {}

            if filters.query:
                filter_obj["or"] = [
                    {"title": {"containsIgnoreCase": filters.query}},
                    {"description": {"containsIgnoreCase": filters.query}},
                ]

            if filters.team:
                teams = await self.get_teams()
                team = next(
                    (
                        t
                        for t in teams
                        if t["key"].lower() == filters.team.lower()
                        or t["name"].lower() == filters.team.lower()
                    ),
                    None,
                )
                if team:
                    filter_obj["team"] = {"id": {"eq": team["id"]}}

            if filters.priority:
                filter_obj["priority"] = {"number": {"in": filters.priority}}

            if filters.state:
                filter_obj["state"] = {"name": {"in": filters.state}}

            variables = {
                "first": filters.limit or 50,
                "filter": filter_obj if filter_obj else None,
            }

            data = await self._execute_query(ISSUES_QUERY, variables)
            issues = data.get("issues", {}).get("nodes", [])

            self._log_activity(
                "search_issues",
                "issue",
                "search",
                {"filters": filters.__dict__, "count": len(issues)},
                True,
            )

            return issues  # type: ignore
        except Exception as e:
            self._log_activity(
                "search_issues", "issue", "search", {"filters": filters.__dict__}, False, str(e)
            )
            raise LinearAPIError(f"Failed to search issues: {e}")

    async def create_comment(self, input_data: LinearCommentInput) -> CommentData:
        try:
            variables = {
                "input": {
                    "body": input_data.body,
                    "issueId": input_data.issue_id,
                }
            }

            data = await self._execute_query(CREATE_COMMENT_MUTATION, variables)
            result = data.get("commentCreate", {})

            if not result.get("success"):
                raise LinearAPIError("Failed to create comment")

            comment = result.get("comment", {})
            self._log_activity(
                "create_comment",
                "comment",
                comment.get("id", "new"),
                {"issueId": input_data.issue_id, "bodyLength": len(input_data.body)},
                True,
            )

            return comment  # type: ignore
        except Exception as e:
            self._log_activity(
                "create_comment", "comment", "new", {"issueId": input_data.issue_id}, False, str(e)
            )
            raise LinearAPIError(f"Failed to create comment: {e}")

    async def get_projects(self, team_id: str | None = None) -> list[ProjectData]:
        try:
            data = await self._execute_query(PROJECTS_QUERY, {"first": 100})
            projects = data.get("projects", {}).get("nodes", [])

            if team_id:
                projects = [
                    p
                    for p in projects
                    if any(t["id"] == team_id for t in p.get("teams", {}).get("nodes", []))
                ]

            self._log_activity(
                "list_projects", "project", "all", {"count": len(projects), "teamId": team_id}, True
            )

            return projects  # type: ignore
        except Exception as e:
            self._log_activity(
                "list_projects", "project", "all", {"teamId": team_id}, False, str(e)
            )
            raise LinearAPIError(f"Failed to fetch projects: {e}")

    async def get_project(self, project_id: str) -> ProjectData:
        try:
            projects = await self.get_projects()
            project = next((p for p in projects if p["id"] == project_id), None)

            if not project:
                raise LinearAPIError(f"Project {project_id} not found")

            self._log_activity(
                "get_project", "project", project_id, {"name": project["name"]}, True
            )

            return project
        except Exception as e:
            self._log_activity("get_project", "project", project_id, {}, False, str(e))
            raise LinearAPIError(f"Failed to fetch project: {e}")

    async def get_users(self) -> list[UserData]:
        try:
            data = await self._execute_query(USERS_QUERY)
            users = data.get("users", {}).get("nodes", [])
            self._log_activity("list_users", "user", "all", {"count": len(users)}, True)
            return users  # type: ignore
        except Exception as e:
            self._log_activity("list_users", "user", "all", {}, False, str(e))
            raise LinearAPIError(f"Failed to fetch users: {e}")

    async def get_current_user(self) -> UserData:
        try:
            data = await self._execute_query(VIEWER_QUERY)
            user = data.get("viewer", {})
            self._log_activity(
                "get_current_user",
                "user",
                user.get("id", "current"),
                {"email": user.get("email"), "name": user.get("name")},
                True,
            )
            return user  # type: ignore
        except Exception as e:
            self._log_activity("get_current_user", "user", "current", {}, False, str(e))
            raise LinearAPIError(f"Failed to fetch current user: {e}")

    async def get_labels(self, team_id: str | None = None) -> list[LabelData]:
        try:
            filter_obj = None
            if team_id:
                filter_obj = {"team": {"id": {"eq": team_id}}}

            data = await self._execute_query(LABELS_QUERY, {"first": 100, "filter": filter_obj})
            labels = data.get("issueLabels", {}).get("nodes", [])

            self._log_activity(
                "list_labels", "label", "all", {"count": len(labels), "teamId": team_id}, True
            )

            return labels  # type: ignore
        except Exception as e:
            self._log_activity("list_labels", "label", "all", {"teamId": team_id}, False, str(e))
            raise LinearAPIError(f"Failed to fetch labels: {e}")

    async def get_workflow_states(self, team_id: str) -> list[StateData]:
        try:
            data = await self._execute_query(
                WORKFLOW_STATES_QUERY, {"filter": {"team": {"id": {"eq": team_id}}}}
            )
            states = data.get("workflowStates", {}).get("nodes", [])

            self._log_activity(
                "list_workflow_states", "team", team_id, {"count": len(states)}, True
            )

            return states  # type: ignore
        except Exception as e:
            self._log_activity("list_workflow_states", "team", team_id, {}, False, str(e))
            raise LinearAPIError(f"Failed to fetch workflow states: {e}")

File: plugins/plugin-linear/typescript/src/services/linear.ts
import { type IAgentRuntime, logger, Service } from "@elizaos/core";
import {
  type Comment,
  type Issue,
  type IssueLabel,
  LinearClient,
  type Project,
  type Team,
  type User,
  type WorkflowState,
} from "@linear/sdk";
import type {
  ActivityDetailObject,
  ActivityDetailValue,
  LinearActivityItem,
  LinearCommentInput,
  LinearConfig,
  LinearIssueInput,
  LinearSearchFilters,
} from "../types";
import { LinearAuthenticationError } from "../types";

export class LinearService extends Service {
  static serviceType = "linear";

  capabilityDescription =
    "Linear API integration for issue tracking, project management, and team collaboration";

  private client: LinearClient;
  private activityLog: LinearActivityItem[] = [];
  private linearConfig: LinearConfig;
  public workspaceId?: string;

  constructor(runtime?: IAgentRuntime) {
    super(runtime);

    const apiKey = runtime?.getSetting("LINEAR_API_KEY") as string;
    const workspaceId = runtime?.getSetting("LINEAR_WORKSPACE_ID") as string;

    if (!apiKey) {
      throw new LinearAuthenticationError("Linear API key is required");
    }

    this.linearConfig = {
      LINEAR_API_KEY: apiKey,
      LINEAR_WORKSPACE_ID: workspaceId,
    };

    this.workspaceId = workspaceId;

    this.client = new LinearClient({
      apiKey: this.linearConfig.LINEAR_API_KEY,
    });
  }

  static async start(runtime: IAgentRuntime): Promise<LinearService> {
    const service = new LinearService(runtime);
    await service.validateConnection();
    logger.info("Linear service started successfully");
    return service;
  }

  async stop(): Promise<void> {
    this.activityLog = [];
    logger.info("Linear service stopped");
  }

  private async validateConnection(): Promise<void> {
    try {
      const viewer = await this.client.viewer;
      logger.info(`Linear connected as user: ${viewer.email}`);
    } catch (_error) {
      throw new LinearAuthenticationError("Failed to authenticate with Linear API");
    }
  }

  private logActivity(
    action: string,
    resourceType: LinearActivityItem["resource_type"],
    resourceId: string,
    details: Record<string, ActivityDetailValue>,
    success: boolean,
    error?: string
  ): void {
    const activity: LinearActivityItem = {
      id: `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`,
      timestamp: new Date().toISOString(),
      action,
      resource_type: resourceType,
      resource_id: resourceId,
      details,
      success,
      error,
    };

    this.activityLog.push(activity);

    if (this.activityLog.length > 1000) {
      this.activityLog = this.activityLog.slice(-1000);
    }
  }

  getActivityLog(limit?: number, filter?: Partial<LinearActivityItem>): LinearActivityItem[] {
    let filtered = [...this.activityLog];

    if (filter) {
      filtered = filtered.filter((item) => {
        return Object.entries(filter).every(([key, value]) => {
          return item[key as keyof LinearActivityItem] === value;
        });
      });
    }

    return filtered.slice(-(limit || 100));
  }

  clearActivityLog(): void {
    this.activityLog = [];
    logger.info("Linear activity log cleared");
  }

  async getTeams(): Promise<Team[]> {
    const teams = await this.client.teams();
    const teamList = await teams.nodes;

    this.logActivity("list_teams", "team", "all", { count: teamList.length }, true);
    return teamList;
  }

  async getTeam(teamId: string): Promise<Team> {
    const team = await this.client.team(teamId);
    this.logActivity("get_team", "team", teamId, { name: team.name }, true);
    return team;
  }

  async createIssue(input: LinearIssueInput): Promise<Issue> {
    const issuePayload = await this.client.createIssue({
      title: input.title,
      description: input.description,
      teamId: input.teamId,
      priority: input.priority,
      assigneeId: input.assigneeId,
      labelIds: input.labelIds,
      projectId: input.projectId,
      stateId: input.stateId,
      estimate: input.estimate,
      dueDate: input.dueDate,
    });

    const issue = await issuePayload.issue;
    if (!issue) {
      throw new Error("Failed to create issue");
    }

    this.logActivity(
      "create_issue",
      "issue",
      issue.id,
      {
        title: input.title,
        teamId: input.teamId,
      },
      true
    );

    return issue;
  }

  async getIssue(issueId: string): Promise<Issue> {
    const issue = await this.client.issue(issueId);
    this.logActivity(
      "get_issue",
      "issue",
      issueId,
      {
        title: issue.title,
        identifier: issue.identifier,
      },
      true
    );
    return issue;
  }

  async updateIssue(issueId: string, updates: Partial<LinearIssueInput>): Promise<Issue> {
    const updatePayload = await this.client.updateIssue(issueId, {
      title: updates.title,
      description: updates.description,
      priority: updates.priority,
      assigneeId: updates.assigneeId,
      labelIds: updates.labelIds,
      projectId: updates.projectId,
      stateId: updates.stateId,
      estimate: updates.estimate,
      dueDate: updates.dueDate,
    });

    const issue = await updatePayload.issue;
    if (!issue) {
      throw new Error("Failed to update issue");
    }

    this.logActivity("update_issue", "issue", issueId, updates, true);
    return issue;
  }

  async deleteIssue(issueId: string): Promise<void> {
    const archivePayload = await this.client.archiveIssue(issueId);

    const success = await archivePayload.success;
    if (!success) {
      throw new Error("Failed to archive issue");
    }

    this.logActivity("delete_issue", "issue", issueId, { action: "archived" }, true);
  }

  async searchIssues(filters: LinearSearchFilters): Promise<Issue[]> {
    const filterObject: Record<string, string | number | boolean | object | null | undefined> = {};

    if (filters.query) {
      filterObject.or = [
        { title: { containsIgnoreCase: filters.query } },
        { description: { containsIgnoreCase: filters.query } },
      ];
    }

    if (filters.team) {
      const teams = await this.getTeams();
      const team = teams.find(
        (t) =>
          t.key.toLowerCase() === filters.team?.toLowerCase() ||
          t.name.toLowerCase() === filters.team?.toLowerCase()
      );

      if (team) {
        filterObject.team = { id: { eq: team.id } };
      }
    }

    if (filters.assignee && filters.assignee.length > 0) {
      const users = await this.getUsers();
      const assigneeIds = filters.assignee
        .map((assigneeName) => {
          const user = users.find(
            (u) =>
              u.email === assigneeName || u.name.toLowerCase().includes(assigneeName.toLowerCase())
          );
          return user?.id;
        })
        .filter(Boolean);

      if (assigneeIds.length > 0) {
        filterObject.assignee = { id: { in: assigneeIds } };
      }
    }

    if (filters.priority && filters.priority.length > 0) {
      filterObject.priority = { number: { in: filters.priority } };
    }

    // Add state filter
    if (filters.state && filters.state.length > 0) {
      filterObject.state = {
        name: { in: filters.state },
      };
    }

    if (filters.label && filters.label.length > 0) {
      filterObject.labels = {
        some: {
          name: { in: filters.label },
        },
      };
    }

    const query = this.client.issues({
      first: filters.limit || 50,
      filter: Object.keys(filterObject).length > 0 ? filterObject : undefined,
    });

    const issues = await query;
    const issueList = await issues.nodes;

    this.logActivity(
      "search_issues",
      "issue",
      "search",
      {
        filters: { ...filters } as ActivityDetailObject,
        count: issueList.length,
      },
      true
    );

    return issueList;
  }

  async createComment(input: LinearCommentInput): Promise<Comment> {
    const commentPayload = await this.client.createComment({
      body: input.body,
      issueId: input.issueId,
    });

    const comment = await commentPayload.comment;
    if (!comment) {
      throw new Error("Failed to create comment");
    }

    this.logActivity(
      "create_comment",
      "comment",
      comment.id,
      {
        issueId: input.issueId,
        bodyLength: input.body.length,
      },
      true
    );

    return comment;
  }

  async getProjects(teamId?: string): Promise<Project[]> {
    // Linear SDK v51 requires manual team filtering on projects
    const query = this.client.projects({
      first: 100,
    });

    const projects = await query;
    let projectList = await projects.nodes;

    if (teamId) {
      const filteredProjects = await Promise.all(
        projectList.map(async (project) => {
          const projectTeams = await project.teams();
          const teamsList = await projectTeams.nodes;
          const hasTeam = teamsList.some((team: Team) => team.id === teamId);
          return hasTeam ? project : null;
        })
      );
      projectList = filteredProjects.filter(Boolean) as Project[];
    }

    this.logActivity(
      "list_projects",
      "project",
      "all",
      {
        count: projectList.length,
        teamId,
      },
      true
    );

    return projectList;
  }

  async getProject(projectId: string): Promise<Project> {
    const project = await this.client.project(projectId);
    this.logActivity(
      "get_project",
      "project",
      projectId,
      {
        name: project.name,
      },
      true
    );
    return project;
  }

  async getUsers(): Promise<User[]> {
    const users = await this.client.users();
    const userList = await users.nodes;

    this.logActivity(
      "list_users",
      "user",
      "all",
      {
        count: userList.length,
      },
      true
    );

    return userList;
  }

  async getCurrentUser(): Promise<User> {
    const user = await this.client.viewer;
    this.logActivity(
      "get_current_user",
      "user",
      user.id,
      {
        email: user.email,
        name: user.name,
      },
      true
    );
    return user;
  }

  async getUserTeams(): Promise<Team[]> {
    const viewer = await this.client.viewer;
    const teams = await viewer.teams();
    const teamList = await teams.nodes;

    this.logActivity(
      "list_user_teams",
      "team",
      viewer.id,
      {
        count: teamList.length,
      },
      true
    );

    return teamList;
  }

  async getLabels(teamId?: string): Promise<IssueLabel[]> {
    const query = this.client.issueLabels({
      first: 100,
      filter: teamId
        ? {
            team: { id: { eq: teamId } },
          }
        : undefined,
    });

    const labels = await query;
    const labelList = await labels.nodes;

    this.logActivity(
      "list_labels",
      "label",
      "all",
      {
        count: labelList.length,
        teamId,
      },
      true
    );

    return labelList;
  }

  async getWorkflowStates(teamId: string): Promise<WorkflowState[]> {
    const states = await this.client.workflowStates({
      filter: {
        team: { id: { eq: teamId } },
      },
    });

    const stateList = await states.nodes;

    this.logActivity(
      "list_workflow_states",
      "team",
      teamId,
      {
        count: stateList.length,
      },
      true
    );

    return stateList;
  }
}
